### PR TITLE
Added preliminary PHP support.

### PR DESCRIPTION
**Pull request by kiuby88**
PHP support is work-in-progress. It allows deploying of PHP web applications from git repositories and tarball sources. 

Below, we can see a  php deployment example using a git repo.

```
name: HelloWorld
services:
- serviceType: brooklyn.entity.webapp.apache.ApacheServer
 name: Apache Server
 location: localhost
 http_port: 80
 app_git_repo_url: https://github.com/kiuby88/phpHelloWorld.git
```

The next plan shows a php deployment example using a tarball resource

```
name: HelloWorld
services:
- serviceType: brooklyn.entity.webapp.apache.ApacheServer
 name: Apache Server
 location: localhost
 http_port: 80
 app_tarball_url : https://github.com/kiuby88/phpHelloWorld.git 
```

Known issues:
- This implementation deploys an Apache Web Server. No application server support yet. 
- Metrics will be provided by Apache mod_status module.
- No HTTPS support yet.

### CODE REVIEW

File: software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java [hunk 1]
```diff
@@ -740,4 +740,72 @@ protected ScriptHelper newScript(Map<String, ?> flags, String phase) {
     @Override
     public void setup() { }
 
+
```
Comment (andreaturli) [inline software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:743]: General comment: I don't think the `AbstractSoftwareProcessSshDriver` needs to install always git. Can you move this code down to `PhpWebAppSshDriver` maybe?

File: software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java [hunk 2]
```diff
@@ -740,4 +740,72 @@ protected ScriptHelper newScript(Map<String, ?> flags, String phase) {
...
+                .add(String.format("git clone %s %s", url, targetDir))
+                .build();
+        log.info("Coping git repository url: {} to {}", new Object[]{url, targetDir});
+        int result= newScript(CUSTOMIZING)
+                .body.append(commands)
+                .execute();
+        return result;
+    }
+
+    private void checkAndInstallGit(){
```
Comment (andreaturli) [inline software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:788]: I'd move that logic to PhpWebAppSshDriver.install

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+import brooklyn.location.access.BrooklynAccessUtils;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Sets;
+import com.google.common.net.HostAndPort;
+import org.slf4j.LoggerFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: try to avoid wildcards

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcess.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public static final AttributeSensor<Set<String>> DEPLOYED_PHP_APPS = new BasicAttributeSensor(
+            Set.class, "webapp.deployedApps", "Names of archives/contexts that are currently deployed");
+    public static final MethodEffector<Void> DEPLOY_GIT_RESOURCE = new MethodEffector<Void>(PhpWebAppSoftwareProcess.class, "deployGitResource");
+    public static final MethodEffector<Void> DEPLOY_TARBALL_RESOURCE = new MethodEffector<Void>(PhpWebAppSoftwareProcess.class, "deployTarballResource");
+    public static final MethodEffector<Void> UNDEPLOY = new MethodEffector<Void>(PhpWebAppSoftwareProcess.class, "undeploy");
+
+    ConfigKey<String> SUGGESTED_VERSION= ConfigKeys.newConfigKeyWithDefault(SoftwareProcess.SUGGESTED_VERSION, "5");
+
+    /**
+     * So, the download method have to define the path.
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcess.java]: Could you rephrase this comment?

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        String gitRepoUrl = getConfig(APP_GIT_REPO_URL);
+        if (gitRepoUrl!=null){
+            String targetName=inferCorrectAppGitName();
+            deployGitResource(gitRepoUrl, targetName);
+        }
+    }
+
+    private String inferCorrectAppGitName(){
+        String result;
+        if(getConfig(APP_NAME)==null || getConfig(APP_NAME).equals("")){
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: Maybe better `Strings.isEmpty(getConfig(APP_NAME))` ?

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        String tarballResourceUrl = getConfig(APP_TARBALL_URL);
+        if (tarballResourceUrl!=null){
+            String targetName=inferCorrectAppTarballName();
+            deployTarballResource(tarballResourceUrl, targetName);
+        }
+    }
+
+    private String  inferCorrectAppTarballName(){
+        String result;
+        if(getConfig(APP_NAME)==null || getConfig(APP_NAME).equals("")){
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: Maybe better `if (Strings.isEmpty(getConfig(APP_NAME)))` ?

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            result=getDriver().getSourceNameResolver().getIdOfTarballFromUrl(getConfig(APP_TARBALL_URL));
+        }
+        else{
+            result=getConfig(APP_NAME);
+        }
+        return result;
+    }
+
+    @Effector(description="Deploys the given artifact, from a source URL, to a given deployment filename/context")
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: move the `@Effector` annotation up to the `PhpWebAppSoftwareProcess` interface

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    private void updateDeploymentSensorToDeployAnApp(String deployedAppName){
+        Set<String> deployedPhpApps = getDeployedPhpAppsAttribute();
+        if (deployedPhpApps == null) {
+            deployedPhpApps = Sets.newLinkedHashSet();
+        }
+        deployedPhpApps.add(deployedAppName);
+        setDeployedPhpAppsAttribute(deployedPhpApps);
+    }
+
+    @Effector(description="Deploys the given artifact, from a source URL, to a given deployment filename/context")
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: same here- move to the interface the annotation

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSshDriver.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+    private int extractTarballResource(String  deployTargetDir, String tarballResourceName){
+        String extractCommand=String.format("tar xzfv %s%s -C %s",deployTargetDir, tarballResourceName, deployTargetDir);
+        return getMachine().execCommands("extract tarball resource", ImmutableList.of(extractCommand));
+    }
+
+    @Override
+    public void install() {
+        //Fixme use the newScript to install php
+//        List<String> commands = ImmutableList.<String>builder()
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSshDriver.java:155]: consider to move here the commands to install git

File: software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+import brooklyn.entity.webapp.PhpWebAppSoftwareProcess;
+import brooklyn.event.AttributeSensor;
+import brooklyn.event.basic.BasicAttributeSensorAndConfigKey.StringAttributeSensorAndConfigKey;
+import brooklyn.event.basic.BasicAttributeSensorAndConfigKey;
+import brooklyn.event.basic.BasicConfigKey;
+import brooklyn.event.basic.PortAttributeSensorAndConfigKey;
+import brooklyn.event.basic.Sensors;
+import brooklyn.util.flags.SetFromFlag;
+
+@Catalog(name="Apache Web Server", description="ApacheServer:  is a web server application which can be connect" +
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java]: Similarly to `Tomcat` entity, I think we should name it 

`@Catalog(name="HTTP Server", description="Apache HTTP Server Project is an open-source HTTP server")` 

as described in the official website http://httpd.apache.org/

File: software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        "with ", iconUrl="classpath:///jboss-logo.png")
+@ImplementedBy(ApacheServerImpl.class)
+public interface ApacheServer extends PhpWebAppSoftwareProcess, PhpWebAppService, HasShortName {
```
Comment (andreaturli) [inline software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java:22]: I'm not completely sure HttpServer should always extend PhpWebApp*
Unless it is a customized entity for Seaclouds, we should make the PHP support optional

@ahgittin what do you think?

File: software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java (hunk 1 above)
Comment (kiuby88) [inline software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:743]: Yea, but I added this method to `AbstractSoftwareProcessSshDriver` as a general funtionality and if a different entity needed install git, it would not need re-write this code. What do you think?

File: software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java (hunk 2 above)
Comment (kiuby88) [inline software/base/src/main/java/brooklyn/entity/basic/AbstractSoftwareProcessSshDriver.java:788]: Same here, this method provides how download  (clone) a git repo using a target Folder in a generic way. All entity that need clone a git repo are able to use this method.

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java (hunk 3 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: I think so, thx.

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcess.java (hunk 4 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcess.java]: This comment and the next are re-written.

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java (hunk 5 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: done

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java (hunk 6 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: done

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java (hunk 7 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: The annotations are duplicate, so, I am going to remove this copy.

File: software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java (hunk 8 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/PhpWebAppSoftwareProcessImpl.java]: Same above

File: software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java (hunk 11 above)
Comment (kiuby88) [inline software/webapp/src/main/java/brooklyn/entity/webapp/apache/ApacheServer.java:22]: @andreaturli 
I defined this architecture because I thought that it could be extensible. For example, for deploy php using JBoss only could be necessary implements a JBoss which will extend PhpWebApp*. However, currently I think that it is not the best point of view, Apache should be an independent entity which should be used for PhpWebApp for deploying the artifacts.
So, I think your question is a good idea and I would like know your thinking about this.
